src/V3EmitMk.cpp

```diff
@@ -24,18 +24,514 @@
+                for (const auto& f : entry.m_concatenatedFilenames) { os << "| + " << f << endl; }
+            } else {
+                os << "+ " << entry.m_fileName << endl;
+            }
+        }
+    }
+
+    void assertFilesSame() const {
+        auto ifIt = m_inputFiles.begin();
+        auto ofIt = m_outputFiles.begin();
+        while (ifIt != m_inputFiles.end() && ofIt != m_outputFiles.end()) {
+            if (ofIt->isConcatenatingFile()) {
+                for (const auto& ocf : ofIt->m_concatenatedFilenames) {
+                    UASSERT(ifIt != m_inputFiles.end(),
+                            "More output files than input files. First extra file: " << ocf);
+                    UASSERT(ifIt->m_filename == ocf,
+                            "Name mismatch: " << ifIt->m_filename << " != " << ocf);
+                    ++ifIt;
+                }
+                ++ofIt;
+            } else {
+                UASSERT(ifIt->m_filename == ofIt->m_fileName,
+                        "Name mismatch: " << ifIt->m_filename << " != " << ofIt->m_fileName);
+                ++ifIt;
+                ++ofIt;
+            }
+        }
+        UASSERT(ifIt == m_inputFiles.end(),
+                "More input files than input files. First extra file: " << ifIt->m_filename);
+        UASSERT(ofIt == m_outputFiles.end(),
+                "More output files than input files. First extra file: " << ofIt->m_fileName);
+    }
+
+public:
+    static std::vector<FileOrConcatenatedFilesList>
+    singleConcatenatedFilesList(std::vector<FilenameWithScore> inputFiles, uint64_t totalScore,
+                                std::string concatenatingFilePrefix) {
+        OutputGroup group{std::move(inputFiles), totalScore, concatenatingFilePrefix};
+        group.process();
+        return group.m_outputFiles;
+    }
+
+    // Called when the concatenation is aborted, creates an identity mapping
+    bool fallbackNoGrouping(size_t inputFilesCount, int totalBucketsNum) {
+        // Return early if there's nothing to do.
+        bool groupingRedundant = false;
+        if (inputFilesCount < MIN_FILES_COUNT) {
+            UINFO(4, "File concatenation skipped: Too few files ("
+                         << m_inputFiles.size() << " < " << MIN_FILES_COUNT << ")" << endl);
+            groupingRedundant = true;
+        }
+        if (inputFilesCount < (MIN_FILES_PER_BUCKET * totalBucketsNum)) {
+            UINFO(4, "File concatenation skipped: Too few files per bucket ("
+                         << m_inputFiles.size() << " < " << MIN_FILES_PER_BUCKET << " - "
+                         << totalBucketsNum << ")" << endl);
+            groupingRedundant = true;
+        }
+        if (!groupingRedundant) return false;
+
+        m_outputFiles.reserve(m_inputFiles.size());
+        for (auto& file : m_inputFiles) {
+            m_outputFiles.push_back({/*fileName=*/std::move(file.m_filename)});
+        }
+        return true;
+    }
+
+    void process() {
```

Comment:
This function is really big, perhaps separate into a few more stage functions and call from here?